pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,349 @@
+	"go.uber.org/atomic"
+)
+
+// consumerGroup is only used to store commit offsets, not for actual consuming.
+const consumerGroup = "mimir"
+
+type record struct {
+	tenantID string
+	content  []byte
+}
+
+type recordConsumer interface {
+	consume(context.Context, []record) error
+}
+
+type PartitionReader struct {
+	services.Service
+	dependencies *services.Manager
+
+	kafkaCfg    KafkaConfig
+	partitionID int32
+
+	client *kgo.Client
+
+	consumer recordConsumer
+	metrics  readerMetrics
+
+	committer      *partitionCommitter
+	commitInterval time.Duration
+
+	logger log.Logger
+}
+
+func NewPartitionReaderForPusher(kafkaCfg KafkaConfig, partitionID int32, pusher Pusher, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newReaderMetrics(partitionID, reg)
+	consumer := newPusherConsumer(pusher, reg, logger)
+	return newPartitionReader(kafkaCfg, partitionID, consumer, logger, metrics)
+}
+
+func newPartitionReader(kafkaCfg KafkaConfig, partitionID int32, consumer recordConsumer, logger log.Logger, metrics readerMetrics) (*PartitionReader, error) {
+	r := &PartitionReader{
+		kafkaCfg:       kafkaCfg,
+		partitionID:    partitionID,
+		consumer:       consumer,
+		metrics:        metrics,
+		commitInterval: time.Second,
+		logger:         log.With(logger, "partition", partitionID),
+	}
+
+	r.Service = services.NewBasicService(r.start, r.run, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffsetWithRetries(ctx)
+	if err != nil {
+		return err
+	}
+	level.Info(r.logger).Log("msg", "resuming consumption from offset", "offset", offset)
+
+	r.client, err = r.newKafkaReader(offset)
+	if err != nil {
+		return errors.Wrap(err, "creating kafka reader client")
+	}
+	r.committer = newConsumerCommitter(r.kafkaCfg, kadm.NewClient(r.client), r.partitionID, r.commitInterval, r.logger)
+
+	r.dependencies, err = services.NewManager(r.committer)
+	if err != nil {
+		return errors.Wrap(err, "creating service manager")
+	}
+	err = services.StartManagerAndAwaitHealthy(ctx, r.dependencies)
+	if err != nil {
+		return errors.Wrap(err, "starting service manager")
+	}
+
+	return nil
+}
+
+func (r *PartitionReader) stop(error) error {
+	level.Info(r.logger).Log("msg", "stopping partition reader")
+
+	err := services.StopManagerAndAwaitStopped(context.Background(), r.dependencies)
+	if err != nil {
+		return errors.Wrap(err, "stopping service manager")
+	}
+	r.client.Close()
+	return nil
+}
+
+func (r *PartitionReader) run(ctx context.Context) error {
+	consumeCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for ctx.Err() == nil {
+		fetches := r.client.PollFetches(ctx)
+		if fetches.Err() != nil {
+			if errors.Is(fetches.Err(), context.Canceled) {
+				return nil
+			}
+			err := collectFetchErrs(fetches)
+			level.Error(r.logger).Log("msg", "encountered error while fetching", "err", err)
+			continue
```

Comment:
What if only some fetches returned error? Aren't we going to loose data because we're not ingesting successful fetches?

Reply:
i opened https://github.com/grafana/mimir/pull/6951 to address these comments